### PR TITLE
mjson routes

### PR DESCRIPTION
**Pull request by sebv**

### CODE REVIEW

File: lib/routes.js [hunk 1]
```diff
@@ -1,17 +1,387 @@
...
+    POST: {command: 'notYetImplemented'}
+  },
+  '/wd/hub/session/:sessionId?/touch/doubleclick': {
+    POST: {command: 'notYetImplemented'}
+  },
+  '/wd/hub/session/:sessionId/touch/longclick': {
+    POST: {command: 'touchLongClick', payloadParams: ['elements']}
+  },
+  '/wd/hub/session/:sessionId/touch/flick': {
+    POST: {command: 'pickAFlickMethod', payloadParams: ['element', 'xspeed', 'yspeed','xoffset', 'yoffset', 'speed']}
```
Comment (jlipps) [inline lib/routes.js]: spacing after ,

File: test/routes-specs.js [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+import 'mochawait';
+
+chai.should();
+
+describe('MJSONWP', () => {
+
+  describe('ensure protocol consistency', () => {
+    it('should not change protocol between patch versions', async () => {
+      var shasum = crypto.createHash('sha1');
+      _(METHOD_MAP).each(function(urlMapping, url) {
```
Comment (jlipps) [inline test/routes-specs.js]:
```
for (let [url, urlMapping] of _.pairs(METHOD_MAP)) {
```

**Comment by moizjv** [comment]
is there a way we can add tests to see we conform to [protocol](https://code.google.com/p/selenium/wiki/JsonWireProtocol) ?

**Comment by jlipps** [comment]
@moizjv that's what his hash test does

File: lib/errors.js [hunk 3]
```diff
@@ -7,6 +7,14 @@ class NotYetImplementedError extends ES6Error {
   }
 }
 
+class NotImplementedError extends ES6Error {
+  constructor () {
+    super("That method is not implemented");
+    this.jsonwpCode = 15; // TODO use status package
```
Comment (jlipps) [inline lib/errors.js]: this should be the same as UnknownError since we are just making this up and it's not part of the spec

Comment (jlipps) [inline lib/errors.js]: actually this is what we are supposed to do with not implemented:

> Every server implementing the WebDriver wire protocol must respond to every defined command. If an individual command has not been implemented on the server, the server should respond with a 501 Not Implemented error message. Note this is the only error in the Invalid Request category that does not return a 4xx status code.

so it shouldn't have a jsonwpCode at all

File: lib/mjsonwp.js [hunk 4]
```diff
@@ -1,12 +1,14 @@
...
 import { validators } from './validators';
-import { NotYetImplementedError, BadParametersError } from './errors';
+import { NotYetImplementedError, NotImplementedError, BadParametersError } from './errors';
 import { METHOD_MAP, NO_SESSION_ID_COMMANDS } from './routes';
 
 class MJSONWP {}
-
 function buildValidatedCommand (command) {
   return async function (...args) {
+    if(!command) {
```
Comment (jlipps) [inline lib/mjsonwp.js]: i don't think we should do this check in this function, but rather below when we are unwrapping the command map. so here:

```js
try {
  if (!spec.command) throw new NotImplementedError();
  checkParams(spec.payloadParams, jsonObj);
```

**Comment by moizjv** [comment]
@jlipps what I meant is was now we are treating this hash as protocol instead of having actual protocol in test spec

**Comment by jlipps** [comment]
let's add a TODO then based on @moizjv's comments

**Comment by sebv** [comment]
We could automatically check it against the spec, but  the spec has changed since we implemented it.

**Pull request merged**